src/test/kotlin/com/terraformation/backend/funder/FundingEntityServiceTest.kt

```diff
@@ -0,0 +1,263 @@
+        )
+    // retrieving both row and model to check createdBy/modifiedBy as well as projects
+    val actualRow = fundingEntitiesDao.fetchOneById(createdModel.id)!!
+    val actualModel = fundingEntityStore.fetchOneById(createdModel.id)
+
+    assertEquals(expected, actualRow)
+    assertEquals(0, actualModel.projects!!.size)
+  }
+
+  @Test
+  fun `create rejects duplicates by name`() {
+    assertThrows<FundingEntityExistsException> { service.create("TestFundingEntity") }
+  }
+
+  @Test
+  fun `create adds project rows`() {
+    insertOrganization()
+    val projectId1 = insertProject()
+    val projectId2 = insertProject()
+    val projectSet = setOf(projectId1, projectId2)
+
+    val inserted = service.create("New Funding Entity with Projects", projectSet)
+
+    val actual = fundingEntityStore.fetchOneById(inserted.id)
+
+    assertEquals(2, actual.projects!!.size)
+    assertEquals(projectSet, actual.projects!!.toSet())
+  }
+
+  @Test
+  fun `update throws exception when no id attached`() {
+    assertThrows<IllegalArgumentException> {
+      service.update(FundingEntitiesRow(name = "Funding Entity without Id"))
+    }
+  }
+
+  @Test
+  fun `update throws exception if user has no permission to manage funding entities`() {
+    every { user.canUpdateFundingEntities() } returns false
+
+    assertThrows<AccessDeniedException> {
+      service.update(FundingEntitiesRow(id = fundingEntityId, name = "Updated Funding Entity"))
+    }
+  }
+
+  @Test
+  fun `update throws exception when name conflict`() {
+    insertFundingEntity("Existing Funding Entity")
+
+    assertThrows<FundingEntityExistsException> {
+      service.update(FundingEntitiesRow(id = fundingEntityId, name = "Existing Funding Entity"))
+    }
+  }
+
+  @Test
+  fun `update throws exception when no matching entity`() {
+    assertThrows<FundingEntityNotFoundException> {
+      service.update(
+          FundingEntitiesRow(id = FundingEntityId(1093), name = "Missing Funding Entity"))
+    }
+  }
+
+  @Test
+  fun `update populates modifiedBy fields`() {
+    val newTime = clock.instant().plusSeconds(1000)
+    clock.instant = newTime
+
+    val newUserId = insertUser()
+
+    val updates =
+        FundingEntitiesRow(
+            id = fundingEntityId,
+            name = "New Name",
+        )
+    val expected =
+        updates.copy(
+            createdBy = user.userId,
+            createdTime = Instant.EPOCH,
+            modifiedBy = newUserId,
+            modifiedTime = newTime,
+        )
+
+    every { user.userId } returns newUserId
+
+    service.update(updates)
+
+    val actual = fundingEntitiesDao.fetchOneById(fundingEntityId)!!
+    assertEquals(expected, actual)
+  }
+
+  @Test
+  fun `update correctly adds new projects`() {
+    val row = FundingEntitiesRow(id = fundingEntityId, name = "TestFundingEntity")
+    val projectId1 = insertProject()
+
+    assertEquals(0, fundingEntityProjectsDao.fetchByFundingEntityId(fundingEntityId).size)
```

Comment:
`DatabaseBackedTest` has a helper method for this: you can use `assertTableEmpty(FUNDING_ENTITY_PROJECTS)`. It will give a more informative error message on failure.

Reply:
Won't this potentially cause concurrency issues if other tests also insert into that table?

Reply:
No, that'll be fine. Each test runs in its own transaction that gets rolled back afterwards. So you don't see the data from other tests when you do `SELECT`.

That might seem like it contradicts what I'm saying about name collisions. The difference is that unique constraints need to be enforced globally, across transactions, even in cases where the colliding row isn't visible in the current transaction. Otherwise we'd get situations where an `INSERT` would claim to have succeeded even though it violated the constraint.